File: Data Structures & Algorithms/sort-an-array/submission-0.cpp
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        quick_sort(nums, 0, nums.size() - 1);
        return nums;
    }

    void quick_sort(std::vector<int>& nums, int l, int r) {
        if (l >= r) return;

        int pivot_index = move(nums, l, r);
        quick_sort(nums, l, pivot_index - 1);
        quick_sort(nums, pivot_index + 1, r);
    }

    int move(std::vector<int>& nums, int l, int r) {
        int pivot = nums[l];

        while (l < r) {
            while (l < r && nums[r] > pivot) r--;
            if (l < r) nums[l++] = nums[r];

            while (l < r && nums[l] < pivot) l++;
            if (l < r) nums[r--] = nums[l];
        }

        nums[l] = pivot;
        return l;
    }
};

File: Data Structures & Algorithms/binary-search/submission-0.cs
public class Solution {
    public int Search(int[] nums, int target) {
        int l = 0, r = nums.Length;
        while (l < r) {
            int middle = (l + r) / 2;
            if (nums[middle] == target) return middle;

            if (nums[middle] > target) r = middle;
            else l = middle + 1;
        }

        return -1;
    }
}
